adr/0014-admission-verify-img-sigs.md

```diff
@@ -0,0 +1,176 @@
+
+
+### Pepr's Homework
+
+- Add a new image verification helper + image verification libraries to the Pepr Controller.
+
+- Expose new functionality as (something like): 
+  ```
+  (registry-ref, public-key-defs) =>  { return bool-if-the-image-sig-is-valid }
+  ```
+
+- New functionality must be able to validate `cosign` image signatures.
+
+
+### Implementation Options
+
+Cosign is a signing & signature verification tool, written and distributed as a Go library and/or standalone binary.
+
+To make it accessible to the Pepr controller one of the following will have to happen:
+
+#### Install CLI
+
+The CLI binary could be shipped along with the Pepr controller & Pepr could just shell out to it.
+
+##### Pros
+
+- Requires minimal new Pepr-custom code
+
+- Easy to install / update
+
+- Simple, well-documented surface area / API
+
+- CLI code from repo under the sigstore umbrella
+
+##### Cons
+
+- All-or-nothing &mdash; can't "shake" out / sub-select unused functionality
+- Deployment size increase - shipping [cosign v2.4.1](https://github.com/sigstore/cosign/releases/download/v2.4.1/cosign-linux-amd64) would add another 105MB to a ~150MB controller image size (nearly 2x).
+
+
+#### Pull-in the Go lib
+
+The Go library could be shipped along with the Pepr controller & Pepr could access it through using a cross-language access technology (i.e. WebAssembly).
+
+##### Pros
+
+- Requires minimal new Pepr-custom code
+
+- Lib code from repo under the sigstore umbrella
+
+- WebAssembly is neat!
+
+##### Cons
+
+- Requires inclusion of a Go lang toolchain f/ development
+
+- Requires install / use an inter-language translation layer (i.e. WebAssembly)
+
+- (Likely) makes runtime debugging more difficult (because .wasm is a binary format )
+
+
+#### Add Go-native Sidecar
+
+The Go library could be shipped as part of a "Go-based microservice" and deployed as a sidecar to the Pepr controller.  Pepr could access it through some form of IPC.
+
+##### Pros
+
+- Makes use of Go-based libraries simple
+
+- As much of the "cloud native" space is built on Go, better positions Pepr for inclusion of other domain-adjacent libs
+
+- Lib code from repo under the sigstore umbrella
+
+
+##### Cons
+
+- Requires inclusion of a Go lang toolchain f/ development
+
+- Requires orchestration between independent processes (i.e. node proc & go proc)
+
+- Requires IPC between independent processes (i.e. node proc & go proc), possibly over a network (i.e. gRPC)
+
+
+#### Install JS-native lib
+
+A JS-native binding / rewrite / wrapper for cosign functionality could be pulled in (i.e. [sigstore-js](https://github.com/sigstore/sigstore-js/tree/main/packages/cli)).
+
+##### Pros
+
+- Written in Pepr-native language (JavaScript) so fits into existing toolset
+
+- Lib code from repo under the sigstore umbrella
+
+##### Cons
+
+- n/a
```

Comment:
We are pretty sure but not 100% yet right?

Reply:
K, put some "hedge" words & a plan of action to validate in there.  Like that better?